feat(linter): implement eslint/no-implicit-coercion rule

tt-a1i opened the pull request:
## Summary

Implements the ESLint `no-implicit-coercion` rule which disallows shorthand type conversions.

### Detected patterns

| Pattern | Type | Suggested Fix |
|---------|------|---------------|
| `!!foo` | boolean | `Boolean(foo)` |
| `+foo` | number | `Number(foo)` |
| `1 * foo` / `foo * 1` | number | `Number(foo)` |
| `foo - 0` | number | `Number(foo)` |
| `- -foo` | number | `Number(foo)` |
| `"" + foo` / `foo + ""` | string | `String(foo)` |
| `~foo.indexOf()` | boolean | - |
| `` `${foo}` `` | string | `String(foo)` |

### Configuration options

```json
{
  "boolean": true,
  "number": true,
  "string": true,
  "disallowTemplateShorthand": false,
  "allow": ["!!", "+"]
}
```

### References

- ESLint rule: https://eslint.org/docs/latest/rules/no-implicit-coercion

## Test plan

- [x] Added comprehensive test cases covering all coercion patterns
- [x] Tests pass locally
- [x] Clippy passes with no warnings

tt-a1i committed:
feat(linter): implement eslint/no-implicit-coercion rule

This rule disallows shorthand type conversions using operators like
`!!`, `+`, `""+ `, `*1`, etc.

Detected patterns:
- `!!foo` → boolean coercion (suggest: Boolean(foo))
- `+foo` → number coercion (suggest: Number(foo))
- `1 * foo` / `foo * 1` → number coercion
- `foo - 0` → number coercion
- `- -foo` → number coercion via double negation
- `"" + foo` / `foo + ""` → string coercion (suggest: String(foo))
- `~foo.indexOf()` → boolean coercion
- \`\${foo}\` → string coercion (when disallowTemplateShorthand is true)

Configuration options:
- boolean: warn on boolean coercion (default: true)
- number: warn on number coercion (default: true)
- string: warn on string coercion (default: true)
- disallowTemplateShorthand: warn on template literal coercion (default: false)
- allow: list of operators to allow (e.g., ["!!", "+"])

tt-a1i committed:
refactor(linter/no-implicit-coercion): improve code quality

- Rename is_number_call to is_already_numeric for clarity
- Extract repeated diagnostic logic into report_coercion helper
- Add CoercionKind enum for type-safe coercion classification
- Improve comments explaining template literal whitespace check

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

camc314 commented on crates/oxc_linter/src/rules/eslint/no_implicit_coercion.rs:
```diff
@@ -0,0 +1,593 @@
+        Expression::NumericLiteral(lit) => (lit.value - 1.0).abs() < f64::EPSILON,
+        _ => false,
+    }
+}
```
you can use `is_number_value` rather than this helper fn;
https://github.com/Boshen/oxc/blob/669afe058011b2c356f16c71c0a4b65dd181c973/crates/oxc_ast/src/ast_impl/js.rs#L137-L140

camc314 reviewed (commented):
This looks good - thanks for working on this!

tt-a1i committed:
refactor(linter): use DefaultRuleConfig and is_number_value

- Use DefaultRuleConfig for simpler configuration parsing
- Replace is_number_one helper with existing Expression::is_number_value

tt-a1i commented:
Thanks for the review! I've addressed all your suggestions:

1. Switched to `DefaultRuleConfig` for cleaner config parsing - good call, it's much simpler! (Note: kept the `Box::new()` wrapper since our struct is `NoImplicitCoercion(Box<...>)`, so can't directly return `into_inner()`)
2. Ran `cargo lintgen` to regenerate `rule_runner_impls.rs`
3. Replaced the `is_number_one` helper with `Expression::is_number_value(1.0)` from oxc_ast

Really appreciate you pointing these out, learned some useful patterns here.

tt-a1i committed:
fix(linter): remove unnecessary hashes in raw string literals

camc314 committed:
tests: add test cases for `lastIndexOf`

eslint doesnt have test cases for these, but it does flag `lastIndexOf`, this commit adds some test cases, and adds support for `lastIndexof`

camc314 committed:
perf(linter/no-implicit-coercion): use bitflags for allow list

Replace `Vec<CompactStr>` with `AllowedOperators` bitflags for checking allowed operators. Since the set of valid operators is fixed and small (6 values), bitflags provide O(1) lookup via bit operations instead of O(n) iteration, and avoid heap allocation for the Vec.

The custom `Deserialize` impl uses a `Visitor` to parse directly from the JSON array without intermediate allocation, and now reports an error for unrecognized operator strings instead of silently ignoring them.

camc314 reviewed (approved):
Thank you! I pushed two more commits, one 

The first adding support for `lastIndexOf`. Annoying I couldn't see any eslint test cases for this scenario. So i added some

https://github.com/eslint/eslint/blob/46eea6d1cbed41d020cb76841ebba30710b0afd0/lib/rules/no-implicit-coercion.js#L14

And the second is refactoring the `allow` config property into bitflags (see the commit message for more detail/why)
